src/oatpp/core/async/CoroutineWaitList.cpp

```diff
@@ -25,16 +26,69 @@
 #include "CoroutineWaitList.hpp"
 
 #include "./Processor.hpp"
+#include <algorithm>
+#include <set>
 
 namespace oatpp { namespace async {
 
 
 CoroutineWaitList::CoroutineWaitList(CoroutineWaitList&& other) {
+  {
+    std::lock_guard<oatpp::concurrency::SpinLock> lock{other.m_lock};
     m_list = std::move(other.m_list);
+  }
+  {
+    std::lock_guard<oatpp::concurrency::SpinLock> lock{other.m_timeoutsLock};
+    m_coroutinesWithTimeout = std::move(other.m_coroutinesWithTimeout);
+    if (!m_coroutinesWithTimeout.empty()) {
+      startTimeoutCheckerThread();
+    }
+  }  
 }
 
 CoroutineWaitList::~CoroutineWaitList() {
   notifyAll();
+  m_stop = true;
+  if (m_thread.joinable()) {
+    m_thread.join();
+  }
+}
+
+void CoroutineWaitList::startTimeoutCheckerThread() {
+  m_thread = std::thread{[this]() { checkCoroutinesForTimeouts(); }};
```

Comment:
Since we don't want to join thread during coroutine destruction I would propose to move this thread to `async::Processor`.
- This thread will serve all the waitlists.
- This thread will call `checkCoroutinesForTimeouts(<processor>)` every `100ms`